lib/ansible/modules/cloud/amazon/s3.py

```diff
@@ -272,131 +240,140 @@
+        # if it's a 404 error, then the object does not exist
+        error_code = int(e.response['Error']['Code'])
+        if error_code == 404:
+            exists = False
+        elif error_code == 403 and validate is False:
+            pass
+        else:
+            module.fail_json(msg="Failed while looking up object (during key check) %s." % obj,
+                             exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+    return exists
+
 
-def keysum(module, s3, bucket, obj, version=None, validate=True):
-    bucket = s3.lookup(bucket, validate=validate)
-    key_check = bucket.get_key(obj, version_id=version)
+def keysum(module, s3, bucket, obj, version=None):
+    if version:
+        key_check = s3.get_object(Bucket=bucket, Key=obj, VersionId=version)
+    else:
+        key_check = s3.get_object(Bucket=bucket, Key=obj)
     if not key_check:
         return None
-    md5_remote = key_check.etag[1:-1]
-    etag_multipart = '-' in md5_remote # Check for multipart, etag is not md5
+    md5_remote = key_check['ETag'][1:-1]
+    etag_multipart = '-' in md5_remote  # Check for multipart, etag is not md5
     if etag_multipart is True:
-        module.fail_json(msg="Files uploaded with multipart of s3 are not supported with checksum, unable to compute checksum.")
+        return None
     return md5_remote
 
+
 def bucket_check(module, s3, bucket, validate=True):
+    exists = True
     try:
-        result = s3.lookup(bucket, validate=validate)
-    except s3.provider.storage_response_error as e:
-        module.fail_json(msg="Failed while looking up bucket (during bucket_check) %s: %s" % (bucket, e),
-                exception=traceback.format_exc())
-    return bool(result)
+        s3.head_bucket(Bucket=bucket)
+    except botocore.exceptions.ClientError as e:
+        # If a client error is thrown, then check that it was a 404 error.
+        # If it was a 404 error, then the bucket does not exist.
+        error_code = int(e.response['Error']['Code'])
+        if error_code == 404:
+            exists = False
+        elif error_code == 403 and validate is False:
+            pass
+        else:
+            module.fail_json(msg="Failed while looking up bucket (during bucket_check) %s." % bucket,
+                             exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+    except botocore.exceptions.EndpointConnectionError as e:
+        module.fail_json(msg="Invalid endpoint provided: %s" % to_text(e), exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+    return exists
+
 
 def create_bucket(module, s3, bucket, location=None):
     if module.check_mode:
         module.exit_json(msg="PUT operation skipped - running in check mode", changed=True)
     if location is None:
-        location = Location.DEFAULT
+        location = 'us-east-1'
+    configuration = {}
+    if location != ('us-east-1' or None):
```

Comment:
This conditional is never going to evaluate properly, since `('us-east-1' or None)` evaluates to `True`, so the only way this evaluates as wrong is if location == True. I think what you mean is `location not in ('us-east-1', None):`